nx/test/nx/random_test.exs

```diff
@@ -0,0 +1,140 @@
+defmodule Nx.RandomTest do
+  use ExUnit.Case, async: true
+
+  import Nx.Helpers, only: [assert_all_close: 2]
+
+  doctest Nx.Random
+
+  describe "key/1" do
+    test "transforms given integer into PRNG key" do
+      key = Nx.Random.key(44)
+
+      assert key |> Nx.type() |> Nx.Type.integer?()
+      assert Nx.shape(key) == {2}
+    end
+  end
+
+  describe "split/2" do
+    test "splits key into multiple keys" do
+      key = Nx.Random.key(33)
+
+      two_keys = Nx.Random.split(key)
+      multiple_keys = Nx.Random.split(key, 12)
+
+      assert Nx.shape(two_keys) == {2, 2}
+      assert Nx.shape(multiple_keys) == {12, 2}
+    end
+  end
+
+  describe "fold_in/2" do
+    test "incorporates integer data into PRNG key" do
+      key = Nx.Random.key(22)
+
+      keys = Enum.map(1..10, &Nx.Random.fold_in(key, &1))
+      assert keys |> Enum.uniq() |> length() == 10
+    end
+
+    test "bigger data" do
+      key = Nx.Random.key(23)
+
+      data = [2 ** 32 - 2, 2 ** 32 - 1]
+      keys = Enum.map(data, &Nx.Random.fold_in(key, &1))
+      assert keys |> Enum.uniq() |> length() == 2
+    end
+  end
+
+  describe "random_bits/2" do
+    test "generates random 32 bit numbers from a key" do
+      key = Nx.Random.key(1701)
+
+      bits = Nx.Random.random_bits(key)
+      expected = Nx.tensor([741_045_208], type: :u32)
+      assert bits == expected
+    end
+
+    test "accepts custom shape" do
+      key = Nx.Random.key(1701)
+
+      bits = Nx.Random.random_bits(key, shape: {3})
+      expected = Nx.tensor([56_197_195, 4_200_222_568, 961_309_823], type: :u32)
+
+      assert bits == expected
+
+      bits = Nx.Random.random_bits(key, shape: {3, 2})
+
+      expected =
+        Nx.tensor(
+          [
+            [927_208_350, 3_916_705_582],
+            [1_835_323_421, 676_898_860],
+            [3_164_047_411, 4_010_691_890]
+          ],
+          type: :u32
+        )
+
+      assert bits == expected
+    end
+  end
+
+  defp to_hex(tensor) do
+    tensor
+    |> Nx.to_flat_list()
+    |> Enum.map(&Integer.to_string(&1, 16))
+  end
+
+  describe "threefry2x32/2" do
```

Comment:
We could add property tests for the uniform and random_integer distributions that generate a sizeable number of samples so we can assert on the values for the first and second moments of the uniform distribution.

For uniform, we can approximate to the continuous case: https://en.wikipedia.org/wiki/Continuous_uniform_distribution#Moments

For random_integer, we can treat it as a discrete uniform distribution: https://mathworld.wolfram.com/DiscreteUniformDistribution.html

(first moment = mean, second moment = variance)

A similar test can be found here (where I needed to test a normal distribution that's generated through a given uniform distribution generator)